### PR TITLE
fix: overflow

### PR DESCRIPTION
**Pull request by paihu**
Sometimes this plugin output very big negative number.
Fix that case (counter overflow).

### CODE REVIEW

**Comment by niku4i** [comment]
Your case must be that you use this plugin with 32bit or 64bit counter(e.g. network devices' interface counter).

The fix looks work for your case, but it does not fit with someone who want to calculate decreasing difference. I intentionally did not support counter overflow because RRD's derive mode does not handle counter over flow.

`min 0` option could be the workaround(but calculated value will be 0 when the tick of overflowing). The another way is we support support RRD's counter mode as plugin's option, like `counter_mode true`.

**Comment by paihu** [comment]
Sorry, I didn't understand derive.

Add counter_mode.

File: lib/fluent/plugin/out_derive.rb [hunk 1]
```diff
@@ -170,6 +171,19 @@ def calc_rate(tag, key, cur_value, prev_value, cur_time, prev_time, adjustment =
     end
   end
 
+  def _calc_rate(cur_value, prev_value)
+    rate = cur_value - prev_value
+    if @counter_mode
+      if rate < 0
+        rate += (1 << 32 - 1)
```
Comment (niku4i) [inline lib/fluent/plugin/out_derive.rb]: This should be `rate += (1 << 32) - 1`

```
irb> (1 << 32 - 1)
=> 2147483648

irb> (1 << 32) - 1
```

File: lib/fluent/plugin/out_derive.rb [hunk 2]
```diff
@@ -170,6 +171,19 @@ def calc_rate(tag, key, cur_value, prev_value, cur_time, prev_time, adjustment =
...
+      end
+      if rate < 0
+        rate += (1 << 64 - 1 << 32 )
```
Comment (niku4i) [inline lib/fluent/plugin/out_derive.rb]: This also returns unwanted value. `-` operator is higher priority than `<<`.

How about `2**32-1` and `2**64 - 2**32`? `**` is higher than `-`.

File: spec/out_derive_spec.rb [hunk 3]
```diff
@@ -299,6 +299,77 @@
...
+          time_unit_division false
+        ]}
+        before do
+          driver.run {
+                  driver.emit({'foo'=> 4294967295}, time)
+                  driver.emit({'foo'=> 100}, time + 60)
+                  }
+        end
+        it {
+                driver.emits[1].should == ['rate', time + 60, {'foo'=> 101}]
```
Comment (niku4i) [inline spec/out_derive_spec.rb]: Thanks for adding test code!! But they are fail, 

```
Failures:

  1) Fluent::DeriveOutput test emit key_pattern counter_mode true 32bit should == ["rate", 1476167965, {"foo"=>101}]
     Failure/Error: driver.emits[1].should == ['rate', time + 60, {'foo'=> 101}]
       expected: ["rate", 1476167965, {"foo"=>101}]
            got: ["rate", 1476167965, {"foo"=>100}] (using ==)
     # ./spec/out_derive_spec.rb:316:in `block (5 levels) in <top (required)>'

  2) Fluent::DeriveOutput test emit key_pattern counter_mode true 64bit should == ["rate", 1476167966, {"foo"=>101}]
     Failure/Error: driver.emits[1].should == ['rate', time + 60, {'foo'=> 101}]
       expected: ["rate", 1476167966, {"foo"=>101}]
            got: ["rate", 1476167966, {"foo"=>100}] (using ==)
     # ./spec/out_derive_spec.rb:351:in `block (5 levels) in <top (required)>'
```

Could you fix the expected values? Then I will merge this.

P.S. the gem spec file `fluent-plugin-derive.gemspec` is old, so you might be not able to run test on your computer.

Comment (paihu) [inline spec/out_derive_spec.rb]: Fix test case and test passed.

Comment (niku4i) [inline spec/out_derive_spec.rb]: Thank you!

**Comment by niku4i** [comment]
Bumped v0.0.4

**Pull request merged**